File: ios/Sources/GearanCore/GearanLink.swift
import Foundation
import CryptoKit

/// Gearan Link Protocol 0.1 framing. Byte-identical to wear/LinkFrameCodec.kt
/// and shared/gearan_link.py. Header 20 B big-endian + HMAC-SHA256 32 B.
public enum GearanLink {
    public static let magic0: UInt8 = 0x47 // "G"
    public static let magic1: UInt8 = 0x52 // "R"
    public static let version: UInt8 = 0x01
    public static let headerLen = 20
    public static let authLen = 32
    public static let maxFramePayload = 512
    public static let maxMessageBytes = 64 * 1024

    public static let serviceUUID = "6E400001-8A21-4A11-9B5C-F3F2A1E4B001"
    public static let rxUUID = "6E400002-8A21-4A11-9B5C-F3F2A1E4B001"
    public static let txUUID = "6E400003-8A21-4A11-9B5C-F3F2A1E4B001"
    public static let ctrlUUID = "6E400004-8A21-4A11-9B5C-F3F2A1E4B001"

    // flags
    public static let flagAckReq: UInt16 = 0x0001
    public static let flagAck: UInt16 = 0x0002
    public static let flagEnc: UInt16 = 0x0004
    public static let flagFrag: UInt16 = 0x0008
    public static let flagRetry: UInt16 = 0x0010
    public static let flagErr: UInt16 = 0x8000

    // message types
    public static let mtAck = 0x01
    public static let mtError = 0x02
    public static let mtPing = 0x03
    public static let mtPong = 0x04
    public static let mtPairHello = 0x05
    public static let mtPairEphemeral = 0x06
    public static let mtPairAuth = 0x07
    public static let mtPairCommit = 0x08
    public static let mtPairConfirm = 0x09
    public static let mtPairAbort = 0x0A
    public static let mtCapsRequest = 0x0B
    public static let mtCapsResponse = 0x0C
    public static let mtSyncStart = 0x0D
    public static let mtSyncData = 0x0E
    public static let mtSyncDone = 0x0F
    public static let mtBattery = 0x10
    public static let mtDeviceInfo = 0x11
    public static let mtNetRequest = 0x12
    public static let mtNetResponse = 0x13
    public static let mtNetError = 0x14
    public static let mtNetCancel = 0x15
    public static let mtNotifPost = 0x16
    public static let mtHealthSample = 0x18
    public static let mtDiagReq = 0x19
    public static let mtDiagResp = 0x1A
    public static let mtDisconnect = 0x1E

    public enum CodecError: Error {
        case tooShort, badMagic, badVersion, badChunks, tooLarge, truncated, authFailure, tooManyConcurrent, chunkMismatch, messageTooLarge
    }

    public struct Frame: Equatable {
        public var messageType: UInt8
        public var flags: UInt16
        public var requestId: UInt32
        public var sequence: UInt32
        public var chunkIndex: UInt16
        public var totalChunks: UInt16
        public var payload: Data
        public init(messageType: UInt8, flags: UInt16, requestId: UInt32, sequence: UInt32, chunkIndex: UInt16, totalChunks: UInt16, payload: Data) {
            self.messageType = messageType; self.flags = flags
            self.requestId = requestId; self.sequence = sequence
            self.chunkIndex = chunkIndex; self.totalChunks = totalChunks
            self.payload = payload
        }
    }

    static func hmac(key: Data, header: Data, payload: Data) -> Data {
        let k = SymmetricKey(data: key)
        var m = HMAC<SHA256>(key: k)
        m.update(data: header); m.update(data: payload)
        return Data(m.finalize())
    }

    public static func encode(_ f: Frame, authKey: Data) throws -> Data {
        guard f.payload.count <= maxFramePayload else { throw CodecError.tooLarge }
        guard f.totalChunks >= 1 else { throw CodecError.badChunks }
        var h = Data(capacity: headerLen)
        h.append(contentsOf: [magic0, magic1, version, f.messageType])
        h.appendUInt16BE(f.flags)
        h.appendUInt32BE(f.requestId)
        h.appendUInt32BE(f.sequence)
        h.appendUInt16BE(f.chunkIndex)
        h.appendUInt16BE(f.totalChunks)
        h.appendUInt16BE(UInt16(f.payload.count))
        let auth = hmac(key: authKey, header: h, payload: f.payload)
        return h + f.payload + auth
    }

    public static func decode(_ data: Data, authKey: Data) throws -> Frame {
        guard data.count >= headerLen + authLen else { throw CodecError.tooShort }
        let h = data.prefix(headerLen)
        let b = [UInt8](h)
        guard b[0] == magic0 && b[1] == magic1 else { throw CodecError.badMagic }
        guard b[2] == version else { throw CodecError.badVersion }
        let type = b[3]
        let flags = (UInt16(b[4]) << 8) | UInt16(b[5])
        let req = (UInt32(b[6]) << 24) | (UInt32(b[7]) << 16) | (UInt32(b[8]) << 8) | UInt32(b[9])
        let seq = (UInt32(b[10]) << 24) | (UInt32(b[11]) << 16) | (UInt32(b[12]) << 8) | UInt32(b[13])
        let cidx = (UInt16(b[14]) << 8) | UInt16(b[15])
        let total = (UInt16(b[16]) << 8) | UInt16(b[17])
        let plen = Int((UInt16(b[18]) << 8) | UInt16(b[19]))
        guard total >= 1 && cidx < total else { throw CodecError.badChunks }
        guard plen <= maxFramePayload else { throw CodecError.tooLarge }
        guard data.count >= headerLen + plen + authLen else { throw CodecError.truncated }
        let payload = data.subdata(in: headerLen..<(headerLen + plen))
        let auth = data.subdata(in: (headerLen + plen)..<(headerLen + plen + authLen))
        let expected = hmac(key: authKey, header: Data(h), payload: payload)
        guard auth == expected else { throw CodecError.authFailure }
        return Frame(messageType: type, flags: flags, requestId: req, sequence: seq, chunkIndex: cidx, totalChunks: total, payload: payload)
    }

    public static func fragment(messageType: UInt8, payload: Data, requestId: UInt32, sequence: UInt32, flags: UInt16 = 0) throws -> [Frame] {
        guard payload.count <= maxMessageBytes else { throw CodecError.messageTooLarge }
        if (payload.isEmpty) { return [Frame(messageType: messageType, flags: flags, requestId: requestId, sequence: sequence, chunkIndex: 0, totalChunks: 1, payload: Data())] }
        var out: [Frame] = []
        let total = (payload.count + maxFramePayload - 1) / maxFramePayload
        for i in 0..<total {
            let s = payload.subdata(in: (i * maxFramePayload)..<min((i + 1) * maxFramePayload, payload.count))
            let f = total > 1 ? (flags | flagFrag) : flags
            out.append(Frame(messageType: messageType, flags: f, requestId: requestId, sequence: sequence, chunkIndex: UInt16(i), totalChunks: UInt16(total), payload: s))
        }
        return out
    }

    /// SAS: 6 digits from SHA256(transcript). Human anti-MITM only, never a key.
    public static func sas(transcript: Data) -> String {
        let d = SHA256.hash(data: transcript)
        let v = (UInt32(d[d.startIndex]) << 24) | (UInt32(d[d.startIndex + 1]) << 16) | (UInt32(d[d.startIndex + 2]) << 8) | UInt32(d[d.startIndex + 3])
        return String(format: "%06d", v % 1_000_000)
    }

    public static func hkdf(ikm: Data, salt: Data, info: Data, length: Int = 32) -> Data {
        let prk = hmac(key: salt, header: ikm, payload: Data())
        var okm = Data(); var t = Data(); var c: UInt8 = 1
        while (okm.count < length) {
            var m = HMAC<SHA256>(key: SymmetricKey(data: prk))
            m.update(data: t); m.update(data: info); m.update(data: Data([c]))
            t = Data(m.finalize()); okm.append(t); c += 1
        }
        return okm.prefix(length)
    }
}

extension Data {
    mutating func appendUInt16BE(_ v: UInt16) { append(UInt8(v >> 8)); append(UInt8(v & 0xFF)) }
    mutating func appendUInt32BE(_ v: UInt32) {
        append(UInt8((v >> 24) & 0xFF)); append(UInt8((v >> 16) & 0xFF))
        append(UInt8((v >> 8) & 0xFF)); append(UInt8(v & 0xFF))
    }
}

/// Sliding replay window (128 entries).
public final class ReplayWindow {
    private let size: UInt32 = 128
    private var maxSeq: Int64 = -1
    private var seen = Set<UInt32>()
    public init() {}
    public func accept(_ seq: UInt32) -> Bool {
        if (seen.contains(seq)) { return false }
        let s = Int64(seq)
        if (s > maxSeq) {
            maxSeq = s; seen.insert(seq)
            let cutoff = maxSeq - Int64(size)
            seen = seen.filter { Int64($0) > cutoff }
            return true
        }
        if (s <= maxSeq - Int64(size)) { return false }
        seen.insert(seq); return true
    }
}

/// Fragment reassembly with timeout + concurrency cap.
public final class Reassembler {
    struct Key: Hashable { var req: UInt32; var type: UInt8 }
    struct Entry { var total: UInt16; var chunks: [UInt16: Data]; var ts: Date }
    private var buffers: [Key: Entry] = [:]
    public var timeout: TimeInterval = 10
    public var maxConcurrent = 8
    public init() {}
    public func feed(_ f: GearanLink.Frame) throws -> Data? {
        buffers = buffers.filter { Date().timeIntervalSince($0.value.ts) <= timeout }
        if (f.totalChunks == 1) { return f.payload }
        let k = Key(req: f.requestId, type: f.messageType)
        var e = buffers[k]
        if (e == nil) {
            guard buffers.count < maxConcurrent else { throw GearanLink.CodecError.tooManyConcurrent }
            e = Entry(total: f.totalChunks, chunks: [:], ts: Date())
            buffers[k] = e
        }
        guard e!.total == f.totalChunks else { throw GearanLink.CodecError.chunkMismatch }
        e!.chunks[f.chunkIndex] = f.payload; e!.ts = Date()
        buffers[k] = e
        if (e!.chunks.count == Int(e!.total)) {
            var out = Data()
            for i in 0..<e!.total { out.append(e!.chunks[i]!) }
            buffers.removeValue(forKey: k)
            return out
        }
        return nil
    }
}

File: shared/gearan_link.py
"""Gearan Link Protocol 0.1 — reference implementation (Python).

Mirrors the framing/auth/chunking/SAS logic implemented natively in
Kotlin (wear/) and Swift (ios/). Used by tests/ on machines without
Android SDK / Xcode.
"""
from __future__ import annotations
import hashlib
import hmac
import os
import struct
import time
from dataclasses import dataclass, field

MAGIC = b"GR"
VERSION = 0x01
HEADER_LEN = 20
MAX_FRAME_PAYLOAD = 512
MAX_MESSAGE_BYTES = 64 * 1024
MAX_TOTAL_CHUNKS = (MAX_MESSAGE_BYTES + MAX_FRAME_PAYLOAD - 1) // MAX_FRAME_PAYLOAD
AUTH_LEN = 32

# flags
FLAG_ACK_REQ = 0x0001
FLAG_ACK = 0x0002
FLAG_ENC = 0x0004
FLAG_FRAG = 0x0008
FLAG_RETRY = 0x0010
FLAG_ERR = 0x8000

# message types
MT_ACK = 0x01
MT_ERROR = 0x02
MT_PING = 0x03
MT_PONG = 0x04
MT_PAIR_HELLO = 0x05
MT_PAIR_EPHEMERAL = 0x06
MT_PAIR_AUTH = 0x07
MT_PAIR_COMMIT = 0x08
MT_PAIR_CONFIRM = 0x09
MT_PAIR_ABORT = 0x0A
MT_CAPS_REQUEST = 0x0B
MT_CAPS_RESPONSE = 0x0C
MT_SYNC_START = 0x0D
MT_SYNC_DATA = 0x0E
MT_SYNC_DONE = 0x0F
MT_BATTERY_UPDATE = 0x10
MT_DEVICE_INFO = 0x11
MT_NETWORK_REQUEST = 0x12
MT_NETWORK_RESPONSE = 0x13
MT_NETWORK_ERROR = 0x14
MT_NETWORK_CANCEL = 0x15

GEARAN_SERVICE_UUID = "6E400001-8A21-4A11-9B5C-F3F2A1E4B001"

HEADER_STRUCT = struct.Struct(">2sBBHIIHHH")  # magic,ver,type,flags,req,seq,chunk,total,len(u16)


@dataclass
class Frame:
    message_type: int
    flags: int
    request_id: int
    sequence: int
    chunk_index: int
    total_chunks: int
    payload: bytes = b""


def compute_auth(header: bytes, payload: bytes, key: bytes) -> bytes:
    return hmac.new(key, header + payload, hashlib.sha256).digest()


def encode_frame(frame: Frame, auth_key: bytes) -> bytes:
    if len(frame.payload) > MAX_FRAME_PAYLOAD:
        raise ValueError("payload exceeds MAX_FRAME_PAYLOAD")
    if (frame.total_chunks < 1 or frame.total_chunks > MAX_TOTAL_CHUNKS
            or frame.chunk_index < 0 or frame.chunk_index >= frame.total_chunks):
        raise ValueError("bad chunk indices")
    header = HEADER_STRUCT.pack(
        MAGIC, VERSION, frame.message_type & 0xFF, frame.flags & 0xFFFF,
        frame.request_id & 0xFFFFFFFF, frame.sequence & 0xFFFFFFFF,
        frame.chunk_index & 0xFFFF, frame.total_chunks & 0xFFFF,
        len(frame.payload),
    )
    auth = compute_auth(header, frame.payload, auth_key)
    return header + frame.payload + auth


def decode_frame(data: bytes, auth_key: bytes) -> Frame:
    if len(data) < HEADER_LEN + AUTH_LEN:
        raise ValueError("frame too short")
    header = data[:HEADER_LEN]
    magic, ver, mtype, flags, req, seq, cidx, total, plen = HEADER_STRUCT.unpack(header)
    if magic != MAGIC:
        raise ValueError("bad magic")
    if ver != VERSION:
        raise ValueError(f"unsupported version {ver}")
    if total == 0 or total > MAX_TOTAL_CHUNKS or cidx >= total:
        raise ValueError("bad chunk indices")
    if plen > MAX_FRAME_PAYLOAD:
        raise ValueError("payload length exceeds max")
    expected_length = HEADER_LEN + plen + AUTH_LEN
    if len(data) < expected_length:
        raise ValueError("truncated frame")
    if len(data) != expected_length:
        raise ValueError("trailing frame data")
    payload = data[HEADER_LEN:HEADER_LEN + plen]
    auth = data[HEADER_LEN + plen:HEADER_LEN + plen + AUTH_LEN]
    expected = compute_auth(header, payload, auth_key)
    if not hmac.compare_digest(auth, expected):
        raise ValueError("authentication failure")
    return Frame(mtype, flags, req, seq, cidx, total, payload)


def fragment_message(message_type: int, payload: bytes, request_id: int,
                     sequence: int, flags: int = 0) -> list[Frame]:
    if len(payload) > MAX_MESSAGE_BYTES:
        raise ValueError("message too large")
    if len(payload) == 0:
        return [Frame(message_type, flags, request_id, sequence, 0, 1, b"")]
    chunks = [payload[i:i + MAX_FRAME_PAYLOAD] for i in range(0, len(payload), MAX_FRAME_PAYLOAD)]
    total = len(chunks)
    out = []
    for idx, c in enumerate(chunks):
        f = flags | (FLAG_FRAG if total > 1 else 0)
        out.append(Frame(message_type, f, request_id, sequence, idx, total, c))
    return out


@dataclass
class Reassembler:
    buffers: dict = field(default_factory=dict)  # (req,type) -> {total, chunks{}, ts}
    timeout_s: float = 10.0
    max_concurrent: int = 8

    def feed(self, frame: Frame) -> bytes | None:
        key = (frame.request_id, frame.message_type)
        now = time.monotonic()
        # expire old
        for k in [k for k, v in self.buffers.items() if now - v["ts"] > self.timeout_s]:
            del self.buffers[k]
        if (frame.total_chunks < 1 or frame.total_chunks > MAX_TOTAL_CHUNKS
                or frame.chunk_index < 0 or frame.chunk_index >= frame.total_chunks
                or len(frame.payload) > MAX_FRAME_PAYLOAD):
            self.buffers.pop(key, None)
            raise ValueError("invalid chunk metadata")
        if frame.total_chunks == 1:
            return bytes(frame.payload)
        entry = self.buffers.get(key)
        if entry is None:
            if len(self.buffers) >= self.max_concurrent:
                raise ValueError("too many concurrent fragmented messages")
            entry = {"total": frame.total_chunks, "chunks": {}, "bytes": 0, "ts": now}
            self.buffers[key] = entry
        if entry["total"] != frame.total_chunks:
            del self.buffers[key]
            raise ValueError("totalChunks mismatch")
        payload = bytes(frame.payload)
        previous = entry["chunks"].get(frame.chunk_index)
        new_size = entry["bytes"] - (len(previous) if previous is not None else 0) + len(payload)
        if new_size > MAX_MESSAGE_BYTES:
            del self.buffers[key]
            raise ValueError("reassembled message too large")
        entry["chunks"][frame.chunk_index] = payload
        entry["bytes"] = new_size
        entry["ts"] = now
        if len(entry["chunks"]) == entry["total"]:
            data = b"".join(entry["chunks"][i] for i in range(entry["total"]))
            del self.buffers[key]
            return data
        return None

    def collect_timeout(self) -> bool:
        now = time.monotonic()
        expired = [k for k, v in self.buffers.items() if now - v["ts"] > self.timeout_s]
        for k in expired:
            del self.buffers[k]
        return bool(expired)


class ReplayWindow:
    """Sliding replay window (last 128 sequences)."""

    def __init__(self, size: int = 128):
        self.size = size
        self.max_seq = -1
        self.seen: set[int] = set()

    def accept(self, seq: int) -> bool:
        if seq in self.seen:
            return False
        if seq > self.max_seq:
            self.max_seq = seq
            self.seen.add(seq)
            # evict old
            cutoff = self.max_seq - self.size
            self.seen = {s for s in self.seen if s > cutoff}
            return True
        if seq <= self.max_seq - self.size:
            return False
        self.seen.add(seq)
        return True


def hkdf_sha256(ikm: bytes, salt: bytes, info: bytes, length: int = 32) -> bytes:
    prk = hmac.new(salt, ikm, hashlib.sha256).digest()
    okm = b""
    t = b""
    counter = 1
    while len(okm) < length:
        t = hmac.new(prk, t + info + bytes([counter]), hashlib.sha256).digest()
        okm += t
        counter += 1
    return okm[:length]


def sas_from_transcript(transcript: bytes) -> str:
    digest = hashlib.sha256(transcript).digest()
    code = int.from_bytes(digest[:4], "big") % 1_000_000
    return f"{code:06d}"


def random_nonce(n: int = 16) -> bytes:
    return os.urandom(n)
